src/historywork/CheckSingleLedgerHeaderWork.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "historywork/CheckSingleLedgerHeaderWork.h"
+#include "history/HistoryManager.h"
+#include "util/Logging.h"
+#include "xdrpp/printer.h"
+
+namespace stellar
+{
+
+CheckSingleLedgerHeaderWork::CheckSingleLedgerHeaderWork(
+    Application& app, std::shared_ptr<HistoryArchive> archive,
+    LedgerHeaderHistoryEntry const& expected)
+    : Work(app, fmt::format("check-single-ledger-header-{}",
+                            expected.header.ledgerSeq))
+    , mCheckpoint(mApp.getHistoryManager().checkpointContainingLedger(
+          expected.header.ledgerSeq))
+    , mArchive(archive)
+    , mExpected(expected)
+{
+}
+
+void
+CheckSingleLedgerHeaderWork::doReset()
+{
+    mGetLedgerFileWork.reset();
+    mDownloadDir =
+        std::make_unique<TmpDir>(mApp.getTmpDirManager().tmpDir(getName()));
+    mFt = std::make_unique<FileTransferInfo>(
```

Comment:
Can only keep it fixed if we don't recreate the download dir, which, well .. again I guess I could modify TmpDir but it seemed fine to just rebuild both, that's the pattern they were designed with.

It's a unique_ptr so it can be default-initialized, without having to construct it redundantly on two separate paths (work constructor & reset). We just fill it in when we reset, which happens at both initial and retry paths of the lifecycle. It could just as well be a std::optional, this was just a lazy pre-C++-17 reflex on my part, I don't much care.

Reply:
(Hah! Just noticed std::option doesn't like forward declarations either. Tradeoffs, tradeoffs.)

Reply:
Well, it definitely shouldn't be optional since it's.. required! Either way, I guess it's fine since you're keeping tmp dir initialization in the reset path as opposed to the constructor. resolving!